### PR TITLE
[RISCV][TTI] Scale the cost of the sext/zext with LMUL

### PR DESCRIPTION
**Pull request by arcbbb**
Use the destination data type to measure the LMUL size for latency/throughput cost

### CODE REVIEW

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 1]
```diff
@@ -915,15 +916,21 @@ InstructionCost RISCVTTIImpl::getCastInstrCost(unsigned Opcode, Type *Dst,
                   (int)Log2_32(Src->getScalarSizeInBits());
     switch (ISD) {
     case ISD::SIGN_EXTEND:
-    case ISD::ZERO_EXTEND:
-      if (Src->getScalarSizeInBits() == 1) {
-        // We do not use vsext/vzext to extend from mask vector.
-        // Instead we use the following instructions to extend from mask vector:
-        // vmv.v.i v8, 0
-        // vmerge.vim v8, v8, -1, v0
```
Comment (lukel97) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:923]: Can we keep the comment?

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 2]
```diff
@@ -915,15 +916,21 @@ InstructionCost RISCVTTIImpl::getCastInstrCost(unsigned Opcode, Type *Dst,
...
-        return 2;
-      }
-      return 1;
+    case ISD::ZERO_EXTEND: {
```
Comment (lukel97) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:919]: There's a FIXME above this that I think we can remove. I don't think we need to worry about the vsetvli toggle either

**Review by lukel97 — APPROVED** [review]
LGTM

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp (hunk 1 above)
Comment (arcbbb) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp]: Sure, somehow I removed it.

**Comment by arcbbb** [comment]
Rebase and address comments.

**Comment by arcbbb** [comment]
reverted since it is failing linux bots https://lab.llvm.org/buildbot/#/builders/5/builds/42101.
```
llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:936:37: runtime error: index -1 out of bounds for type 'unsigned int[3]'
```

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 3]
```diff
@@ -906,23 +906,33 @@ InstructionCost RISCVTTIImpl::getCastInstrCost(unsigned Opcode, Type *Dst,
...
+      return getRISCVInstructionCost({RISCV::VMV_V_I, RISCV::VMERGE_VIM},
+                                     DstLT.second, CostKind);
     }
-    return 1;
+    if (PowDiff > 3)
+      return BaseT::getCastInstrCost(Opcode, Dst, Src, CCH, CostKind, I);
+    unsigned SExtOp[] = {RISCV::VSEXT_VF2, RISCV::VSEXT_VF4, RISCV::VSEXT_VF8};
+    unsigned ZExtOp[] = {RISCV::VZEXT_VF2, RISCV::VZEXT_VF4, RISCV::VZEXT_VF8};
+    unsigned Op =
+        (ISD == ISD::SIGN_EXTEND) ? SExtOp[PowDiff - 1] : ZExtOp[PowDiff - 1];
```
Comment (lukel97) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:933]: I guess something is calling this with a sign/zero extend to the same type, i.e. PowDiff == 0? Probably best just to defensively check for it then

Comment (arcbbb) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:933]: Yes, this failure showed that loop vectorizer `getReductionPatternCost` would call this with extend to the same type.
Will recommit with an additional check on this.Will recommit with an additional check on this.

**Pull request merged**